### PR TITLE
MB-9372 add handler for prime sit extension

### PR DESCRIPTION
**Pull request by eamahanna**
## Description

This PR covers quite a few things:
- Adds handler for createSitExtension for the prime api
- Updates the ghc swagger definition for `sitExtensions` to use the shared definition file
- Updates `getMove` endpoint to include sitExtensions
- Updates the response for the `updateShipment` endpoint to return sitExtensions
- Adds tests

## Reviewer Notes

I am in a bit of a time crunch this morning. I can add more detailed instructions on how to test if need be. Just ping me and let me know. This is a big PR so I wanted to allow people time to review it. I should be able to add instructions after the weekly demo.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
echo "Code goes here"
```

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-9372) for this change

### CODE REVIEW

File: pkg/handlers/ghcapi/mto_shipment_test.go [hunk 1]
```diff
@@ -1604,7 +1604,7 @@ func (suite *HandlerSuite) TestApproveSITExtensionHandler() {
 		suite.IsType(&shipmentops.ApproveSitExtensionOK{}, response)
 		suite.Equal(int64(30), *payload.SitDaysAllowance)
 		suite.Equal("APPROVED", payload.SitExtensions[0].Status)
-		suite.Equal(officeRemarks, payload.SitExtensions[0].OfficeRemarks)
+		suite.Equal(&officeRemarks, payload.SitExtensions[0].OfficeRemarks)
```
Comment (jacquelineIO) [inline pkg/handlers/ghcapi/mto_shipment_test.go:1607]: Why the change from value to reference? Did the structure change on the right-hand side?

File: pkg/handlers/ghcapi/mto_shipment_test.go [hunk 2]
```diff
@@ -1673,7 +1673,7 @@ func (suite *HandlerSuite) CreateSITExtensionAsTOO() {
 		createParams := shipmentops.CreateSitExtensionAsTOOParams{
 			HTTPRequest: req,
 			IfMatch:     eTag,
-			Body: &ghcmessages.CreateSITExtensionAsTOO{
+			Body: &ghcmessages.CreateSitExtensionAsTOO{
```
Comment (jacquelineIO) [inline pkg/handlers/ghcapi/mto_shipment_test.go:1676]: Why are these `Sit` and not `SIT`? (just curious) Maybe it's because it's coming from Swagger.

File: pkg/handlers/primeapi/payloads/model_to_payload_test.go [hunk 3]
```diff
@@ -71,24 +71,24 @@ func (suite *PayloadsSuite) TestReweigh() {
 		UpdatedAt:   updatedAt,
 	}
 
-	suite.T().Run("Success - Returns a basic rewweigh payload without optional fields", func(t *testing.T) {
+	suite.T().Run("Success - Returns a rewweigh payload without optional fields", func(t *testing.T) {
```
Comment (jacquelineIO) [inline pkg/handlers/primeapi/payloads/model_to_payload_test.go:74]: typo `rewweigh`

File: swagger-def/prime.yaml [hunk 4]
```diff
@@ -492,6 +492,52 @@ paths:
           $ref: '#/responses/UnprocessableEntity'
         '500':
           $ref: '#/responses/ServerError'
+  '/mto-shipments/{mtoShipmentID}/sit-extensions':
+    post:
+      summary: createSITExtension
```
Comment (jacquelineIO) [inline swagger-def/prime.yaml:497]: naming style isn't consistent e.g. `CreateSitExtensionAsTOO` under swagger-def/ghc.yaml

File: swagger-def/prime.yaml [hunk 5]
```diff
@@ -492,6 +492,52 @@ paths:
...
+          name: mtoShipmentID
+          description: UUID of the shipment associated with the agent
+          required: true
+          format: uuid
+          type: string
+        - in: body
+          required: true
+          name: body
+          schema:
+            $ref: '#/definitions/CreateSitExtension'
```
Comment (jacquelineIO) [inline swagger-def/prime.yaml:521]: another comment on consistency e.g. `createSITExtension` from this same definition

**Review by jacquelineIO — COMMENTED** [review]
did not review for approval, but had some general comments

File: pkg/handlers/primeapi/move_task_order_test.go [hunk 6]
```diff
@@ -127,6 +127,28 @@ func (suite *HandlerSuite) TestGetMoveTaskOrder() {
...
+		})
+
+		response := handler.Handle(params)
+		suite.IsNotErrResponse(response)
+		suite.IsType(&movetaskorderops.GetMoveTaskOrderOK{}, response)
+
+		moveResponse := response.(*movetaskorderops.GetMoveTaskOrderOK)
+		movePayload := moveResponse.Payload
+		reweighPayload := movePayload.MtoShipments[0].SitExtensions[0]
+		suite.Equal(movePayload.ID.String(), successMove.ID.String())
```
Comment (monfresh) [inline pkg/handlers/primeapi/move_task_order_test.go:148]: The expected and actual are reversed here. It should be
```golang
suite.Equal(successMove.ID.String(), movePayload.ID.String())
```

File: swagger-def/prime.yaml (hunk 5 above)
Comment (eamahanna) [reply swagger-def/prime.yaml]: I just added a commit to change the swagger casings to SITExtension

File: swagger-def/definitions/SitExtension.yaml [hunk 7]
```diff
@@ -0,0 +1,60 @@
+type: object
```
Comment (sandy-wright) [inline swagger-def/definitions/SitExtension.yaml:1]: Thanks for breaking out this definition! Looks like the case on both of the SIT filenames will need to be updated again: `SitExtension(s)` > `SITExtension(s)`

File: pkg/handlers/primeapi/payloads/model_to_payload_test.go [hunk 8]
```diff
@@ -103,16 +103,80 @@ func (suite *PayloadsSuite) TestReweigh() {
...
+		sitExtension.DecisionDate = &decisionDate
+
+		// Send model through func
+		returnedPayload := SITExtension(&sitExtension)
+
+		suite.IsType(&primemessages.SitExtension{}, returnedPayload)
+		suite.Equal(strfmt.UUID(sitExtension.ID.String()), returnedPayload.ID)
+		suite.Equal(strfmt.UUID(sitExtension.MTOShipmentID.String()), returnedPayload.MtoShipmentID)
+		suite.Equal(strfmt.DateTime(returnedPayload.CreatedAt), returnedPayload.CreatedAt)
+		suite.Equal(strfmt.DateTime(returnedPayload.UpdatedAt), returnedPayload.UpdatedAt)
```
[the remark this answers is not in the document]
Comment (eamahanna) [reply pkg/handlers/primeapi/payloads/model_to_payload_test.go:174]: Oops. this was a copy paste error. This is suppose to be not nil

[the remark this answers is not in the document]
Comment (eamahanna) [reply pkg/handlers/primeapi/payloads/model_to_payload_test.go]: fixed

File: pkg/handlers/primeapi/sit_extension_test.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/transcom/mymove/pkg/gen/primemessages"
+	"github.com/transcom/mymove/pkg/handlers"
+	"github.com/transcom/mymove/pkg/models"
+	sitextensionservice "github.com/transcom/mymove/pkg/services/sit_extension"
+	"github.com/transcom/mymove/pkg/testdatagen"
+)
+
+func (suite *HandlerSuite) CreateSITExtensionHandler() {
+	// Make an available MTO
+	mto := testdatagen.MakeAvailableMove(suite.DB())
```
Comment (monfresh) [inline pkg/handlers/primeapi/sit_extension_test.go:21]: `mto` is an old reference we're trying to move away from. We want to eventually replace all `mto`  and `move task order` reference to `move` instead.

File: pkg/handlers/primeapi/sit_extension_test.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+		suite.Equal(remarks, sitExtensionResponse.ContractorRemarks)
+		suite.NotNil(sitExtensionResponse.ID)
+		suite.NotNil(sitExtensionResponse.CreatedAt)
+		suite.NotNil(sitExtensionResponse.UpdatedAt)
+		suite.NotNil(sitExtensionResponse.ETag)
+	})
+
+	suite.T().Run("Failure 402 - Shipment not found", func(t *testing.T) {
+		// Testcase:   Shipment ID is not found
+		// Expected:   Success response 402
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/sit_extension_test.go:82]: I think this should be "422" here. And to clarify, we get a 422 from this endpoint when the shipment isn't found?

File: pkg/handlers/primeapi/sit_extension_test.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	mtoShipment1 := testdatagen.MakeMTOShipment(suite.DB(), testdatagen.Assertions{
+		Move: mto,
+	})
+
+	// Make sit extension
+	daysRequested := int64(30)
+	remarks := "We need an extension"
+	reason := "AWAITING_COMPLETION_OF_RESIDENCE"
+
+	sitExtension := &primemessages.CreateSitExtension{
```
Comment (monfresh) [inline pkg/handlers/primeapi/sit_extension_test.go]: What's the reasoning behind using primemessages to create the extension versus testdatagen?

File: pkg/handlers/primeapi/payloads/model_to_payload_test.go [hunk 12]
```diff
@@ -71,24 +71,24 @@ func (suite *PayloadsSuite) TestReweigh() {
...
-		suite.Equal(strfmt.UUID(returnedPayload.ShipmentID.String()), returnedPayload.ShipmentID)
-		suite.Equal(strfmt.DateTime(returnedPayload.RequestedAt), returnedPayload.RequestedAt)
+		suite.Equal(strfmt.UUID(reweigh.ID.String()), returnedPayload.ID)
+		suite.Equal(strfmt.UUID(reweigh.ShipmentID.String()), returnedPayload.ShipmentID)
+		suite.Equal(strfmt.DateTime(reweigh.RequestedAt), returnedPayload.RequestedAt)
 		suite.Equal(primemessages.ReweighRequester(reweigh.RequestedBy), returnedPayload.RequestedBy)
-		suite.Equal(strfmt.DateTime(returnedPayload.CreatedAt), returnedPayload.CreatedAt)
-		suite.Equal(strfmt.DateTime(returnedPayload.UpdatedAt), returnedPayload.UpdatedAt)
+		suite.Equal(strfmt.DateTime(reweigh.CreatedAt), returnedPayload.CreatedAt)
+		suite.Equal(strfmt.DateTime(reweigh.UpdatedAt), returnedPayload.UpdatedAt)
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/payloads/model_to_payload_test.go:83]: jinx!

File: pkg/handlers/primeapi/sit_extension_test.go (hunk 11 above)
Comment (monfresh) [reply pkg/handlers/primeapi/sit_extension_test.go]: Ah, nevermind. The `CreateSitExtension` threw me off. I thought we were creating an extension in the DB here, but this is just setting up the object.

File: pkg/handlers/ghcapi/mto_shipment_test.go [hunk 13]
```diff
@@ -1727,9 +1727,9 @@ func (suite *HandlerSuite) CreateSITExtensionAsTOO() {
 		suite.NoError(createParams.Body.Validate(strfmt.Default))
 
 		response := handler.Handle(createParams)
-		okResponse := response.(*shipmentops.CreateSitExtensionAsTOOOK)
+		okResponse := response.(*shipmentops.CreateSITExtensionAsTOOOK)
 		payload := okResponse.Payload
-		suite.IsType(&shipmentops.CreateSitExtensionAsTOOOK{}, response)
+		suite.IsType(&shipmentops.CreateSITExtensionAsTOOOK{}, response)
```
Comment (sandy-wright) [inline pkg/handlers/ghcapi/mto_shipment_test.go:1732]: TOOOOOOOOOOOK

File: pkg/handlers/primeapi/sit_extension_test.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+		RequestReason:     &reason,
+	}
+
+	// Create handler
+	handler := CreateSITExtensionHandler{
+		handlers.NewHandlerContext(suite.DB(), suite.TestLogger()),
+		sitextensionservice.NewSitExtensionCreator(),
+	}
+
+	suite.T().Run("Success 201 - Creat sit extension", func(t *testing.T) {
```
Comment (monfresh) [inline pkg/handlers/primeapi/sit_extension_test.go:45]: typo: Create SIT extension

File: pkg/handlers/primeapi/sit_extension_test.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+		suite.Equal(daysRequested, sitExtensionResponse.RequestedDays)
+		suite.Equal(models.SITExtensionRequestReasonAwaitingCompletionOfResidence, sitExtensionResponse.RequestReason)
+		suite.Equal(remarks, sitExtensionResponse.ContractorRemarks)
+		suite.NotNil(sitExtensionResponse.ID)
+		suite.NotNil(sitExtensionResponse.CreatedAt)
+		suite.NotNil(sitExtensionResponse.UpdatedAt)
+		suite.NotNil(sitExtensionResponse.ETag)
+	})
+
+	suite.T().Run("Failure 402 - Shipment not found", func(t *testing.T) {
```
Comment (monfresh) [inline pkg/handlers/primeapi/sit_extension_test.go:80]: Did you mean 404 for not found? This test is returning an UnprocessableEntity error, which would be a 422, but for a bad shipment ID, I would expect a 404, no?

File: pkg/handlers/ghcapi/mto_shipment_test.go (hunk 1 above)
Comment (sandy-wright) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1607]: good catch. Since these are string values, we probably want: 

```go
suite.Equal(officeRemarks, *payload.SitExtensions[0].OfficeRemarks)
```

instead, so it's actually comparing the value of the string and not the pointer address. A slightly more complete test would be:

```go
suite.Require().NotNil(payload.SitExtensions[0].OfficeRemarks) // so the test stops before a potential nil dereference error
suite.Equal(officeRemarks, *payload.SitExtensions[0].OfficeRemarks)
```

File: pkg/handlers/primeapi/sit_extension_test.go (hunk 15 above)
Comment (monfresh) [reply pkg/handlers/primeapi/sit_extension_test.go:80]: I would also add a 422 test for validation errors, such as a missing Request Reason.

**Review by sandy-wright — COMMENTED** [review]
got a few comments, but overall looks good! Just posting the comments now so you can look, but I still need to finish testing the endpoint.

File: pkg/services/sit_extension/sit_extension_creator.go [hunk 16]
```diff
@@ -33,6 +33,9 @@ func (f *sitExtensionCreator) CreateSITExtension(appCtx appcontext.AppContext, s
 		return nil, services.NewNotFoundError(sitExtension.MTOShipmentID, "while looking for MTOShipment")
 	}
 
+	// Set status for sit extension
+	sitExtension.Status = models.SITExtensionStatusPending
```
Comment (monfresh) [inline pkg/services/sit_extension/sit_extension_creator.go:37]: Since this is hardcoding the status in the service function, would it make sense to rename it to `CreatePendingSitExtension` to better communicate what's happening?

File: pkg/handlers/ghcapi/mto_shipment_test.go (hunk 1 above)
Comment (eamahanna) [reply pkg/handlers/ghcapi/mto_shipment_test.go:1607]: Updated!

File: pkg/handlers/ghcapi/mto_shipment_test.go (hunk 2 above)
Comment (eamahanna) [reply pkg/handlers/ghcapi/mto_shipment_test.go]: Yep! they are generated from the swagger definitions.

File: pkg/handlers/primeapi/move_task_order_test.go (hunk 6 above)
Comment (eamahanna) [reply pkg/handlers/primeapi/move_task_order_test.go:148]: fixed!

File: pkg/handlers/primeapi/sit_extension.go [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+// Handle created a sit extension for a shipment
+func (h CreateSITExtensionHandler) Handle(params mtoshipmentops.CreateSITExtensionParams) middleware.Responder {
+	logger := h.LoggerFromRequest(params.HTTPRequest)
+	appCtx := appcontext.NewAppContext(h.DB(), logger)
+
+	// Get the new extension model
+	SITExtension := payloads.SITExtensionModel(params.Body, params.MtoShipmentID)
+
+	// Call the service object
+	createdextension, err := h.SITExtensionCreator.CreateSITExtension(appCtx, SITExtension)
```
[the remark this answers is not in the document]
Comment (eamahanna) [reply pkg/handlers/primeapi/sit_extension.go:31]: sure

File: pkg/handlers/primeapi/sit_extension_test.go (hunk 10 above)
Comment (eamahanna) [reply pkg/handlers/primeapi/sit_extension_test.go:82]: +1 yep this should be a 422.  That was another typo.

File: pkg/services/sit_extension/sit_extension_creator.go (hunk 16 above)
Comment (eamahanna) [reply pkg/services/sit_extension/sit_extension_creator.go:37]: You have a point. I will change this so only if a status is not provided, then it is set to pending.

**Review by sandy-wright — APPROVED** [review]
LGTM! Thanks for making all those changes!
